src/thermal_conductivity/scattering_threephonon.f90

```diff
@@ -0,0 +1,395 @@
+                f1 = perm*psisq*plf1*(lo_gauss(om1, om2 + om3, sigma) - lo_gauss(om1, -om2 - om3, sigma))
+
+                ! And we add everything for each triplet equivalent to the one we are actually computing
+                do i = 1, size(red_triplet, 2)
+                    ! We accumulate the linewidth
+                    g0 = g0 + f0 + f1
+
+                    ! And also the off-diagonal part of the scattering matrix
+                    q2p = red_triplet(1, i)
+                    q3p = red_triplet(2, i)
+                    od_terms(q2p, b2) = od_terms(q2p, b2) + 2.0_r8*(f0 + f1)*om2/om1
+                    od_terms(q3p, b3) = od_terms(q3p, b3) + 2.0_r8*(f0 + f1)*om3/om1
+                end do
+            end do
+        end do
+    end do compute_loop
+
+    ! Now we can symmetrize the off-diagonal contribution
+    ! This can be done in a way to put a value to for mode that have been skipped by the Monte-Carlo !
+    symmetrize_and_distribute: block
+        !> To keep track of the number of mode we actually add
+        integer, dimension(dr%n_mode) :: nn
+        !> To hold the q-point in reduced coordinates
+        real(r8), dimension(3) :: qv2p
+        !> To get the index of the new triplet on the fft_grid
+        integer, dimension(3) :: gi
+        !> Some buffer
+        real(r8), dimension(dr%n_mode) :: buf_xi
+        !> Some integers for the do loops and stuff
+        integer :: j, k, i2
+
+        ! Let's average the off diagonal term
+        allq2: do qi = 1, mcg%npoints
+            q2 = qgridfull(qi)
+            buf_xi = 0.0_r8
+            nn = 0
+            ! First we get the average value
+            do j = 1, qp%ip(q1)%n_invariant_operation
+                k = qp%ip(q1)%invariant_operation(j)
+                ! First we generate q2'
+                select type (qp); type is (lo_fft_mesh)
+                    qv2 = matmul(uc%inv_reciprocal_latticevectors, qp%ap(q2)%r)
+                    qv2p = lo_operate_on_vector(uc%sym%op(k), qv2, reciprocal=.true., fractional=.true.)
+                    if (qp%is_point_on_grid(qv2p) .eqv. .false.) cycle
+                    gi = qp%index_from_coordinate(qv2p)
+                    q2p = qp%gridind2ind(gi(1), gi(2), gi(3)) ! this is R*q'
+                end select
+                ! If q2p < q2, we already did this guy
+                if (q2p .lt. q2) cycle allq2
+                ! Accumulate values for each equivalent term, but avoid those which where not computed
+                do b2 = 1, dr%n_mode
+                    if (od_terms(q2p, b2) .gt. 0.0_r8) then
+                        nn(b2) = nn(b2) + 1
+                        buf_xi(b2) = buf_xi(b2) + od_terms(q2p, b2)
+                    end if
+                end do
+            end do
+            ! And now we can distribute it
+            do j = 1, qp%ip(q1)%n_invariant_operation
+                k = qp%ip(q1)%invariant_operation(j)
+                select type (qp); type is (lo_fft_mesh)
```

Comment:
We need it the first time to take the average and the second time to distribute it.
But it might be possible to only get the equivalent points in a list and then average/distribute in one go. But the way I see it, we would still need to loop two times to be able to allocate the list of equivalent points